lib/spack/spack/concretize.py

```diff
@@ -328,13 +364,29 @@ def concretize_variants(self, spec):
         preferred_variants = PackagePrefs.preferred_variants(spec.name)
         pkg_cls = spec.package_class
         for name, variant in pkg_cls.variants.items():
+            any_set = False
+            var = spec.variants.get(name, None)
+            if var and 'any' in var:
```

Comment:
The `__contains__` operator for variants makes a tuple of the values and checks whether the given value is one of the values. So this check would not match `foo=many`. It matches `foo=any` or `foo=any,another`, although the latter is forbidden by other parts of the code.

Reply:
Is the added support for "any" necessary for this PR?

Reply:
No, but it is necessary for the user to be able to change the projections on the basis of whether a `dev_path` variant is present, regardless of its value. That's not essential to this PR, but it does enable workflows that are otherwise possible for almost any value on the spec to be extended to `dev_path`.

```
projections:
  dev_path=any: 'dev_builds/{name}-{version}-{hash:7}'
  all: '{archictecture}/{compiler.name}-{compiler.version}/{name}-{version}-{hash}'
```

That sort of workflow (changing the value for all direct dev builds but not for transitive dev builds) is impossible in this PR without the `any` support.